newIDE/app/src/ResourcesList/EmbeddedResourceSources.js

```diff
@@ -0,0 +1,142 @@
+import optionalRequire from '../Utils/OptionalRequire';
+import newNameGenerator from '../Utils/NewNameGenerator';
+import { isPathInProjectFolder } from './ResourceUtils';
+import { createNewResource } from './ResourceSource';
+const fs = optionalRequire('fs');
+const path = optionalRequire('path');
+
+/**
+ * Copy the embedded resources inside the project folder
+ * @param project The project
+ * @param embeddedFiles The embedded files
+ * @returns
+ */
+export async function copyEmbeddedToProjectFolder(project, embeddedFiles) {
+  if (!fs || !path) {
+    return;
+  }
+
+  const projectPath = path.dirname(project.getProjectFile());
+  const copies = [];
+
+  for (const { outside, allFiles } of embeddedFiles.values()) {
+    if (outside) {
+      for (const embedded of Object.values(allFiles)) {
+        if (embedded.outside) {
+          const resourceBasename = path.basename(embedded.fullPath);
+          const fileExtension = path.extname(resourceBasename);
+          const fileNameWithoutExtension = path.basename(
+            resourceBasename,
+            fileExtension
+          );
+
+          const newFileNameWithoutExtension = newNameGenerator(
+            fileNameWithoutExtension,
+            tentativeFileName => {
+              const tentativePath =
+                path.join(projectPath, tentativeFileName) + fileExtension;
+              return fs.existsSync(tentativePath);
+            }
+          );
+
+          const resourceNewPath = path.join(
+            projectPath,
+            newFileNameWithoutExtension + fileExtension
+          );
+
+          embedded.resourceName = newFileNameWithoutExtension + fileExtension;
+
+          copies.push(fs.promises.copyFile(embedded.fullPath, resourceNewPath));
+        }
+      }
+    }
+  }
+
+  return Promise.all(copies);
+}
+
+/**
+ * Create the mapping an embedded resource path to its resource name
+ * @param project The project
+ * @param embeddedFiles The embedded files
+ */
+export function mapEmbeddedFiles(project, embeddedFiles) {
+  const projectPath = path.dirname(project.getProjectFile());
+
+  for (const { allFiles, mapping } of embeddedFiles.values()) {
+    for (let { kind, relPath, fullPath, resourceName } of Object.values(
+      allFiles
+    )) {
+      if (!resourceName) {
+        resourceName = path.relative(projectPath, fullPath);
+      }
+
+      const newResource = createNewResource(kind);
+      newResource.setName(resourceName);
+      newResource.setFile(resourceName);
+
+      mapping[relPath] = resourceName;
```

Comment:
I think this is where we could save `{ resourceName, kind }` (or better: `{ resourceName, resourceKind }` to be consistent in our naming, and this make it easier to grep "resourceKind" in the whole codebase rather than just "kind").

Reply:
The function `mapEmbeddedFiles` is used when adding a new resource with embedded resources. The `newResource` is an embedded resource and is `gd.Resource`, I don't think I can change that. The mapping is done with `mapping[relPath] = resourceName`.
But I will rename `kind` as `resourceKind`.

Reply:
I've renamed the variable and with the types for flowjs , the code is clearer (even if it's still a `.js`).